plugins/catalog-backend/src/catalog/CoalescedEntitiesCatalog.ts

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Entity } from '@backstage/catalog-model';
+import { EntityFilters } from '../database';
+import { EntitiesCatalog } from './types';
+
+/**
+ * A simple coalescing catalog wrapper, that acts as a front for collecting
+ * catalog data from multiple sources.
+ *
+ * One possible usage could be to have this as a front to both a
+ * DatabaseEntitiesCatalog that holds Component kinds, and another company-
+ * specific catalog that is a thin wrapper on top of LDAP that supplies Group
+ * and User entities. That way you'll get a coherent view of two very different
+ * entity sources.
+ *
+ * This is mainly meant as a functional example, and you may want to provide
+ * your own more specialized collector if you have this distinct need. This
+ * one does not support adding/updating entities through the API for example.
+ * A more competent implementation may direct the writes to different catalogs
+ * based on entity kind or similar.
+ */
+export class CoalescedEntitiesCatalog implements EntitiesCatalog {
+  private inner: EntitiesCatalog[];
+
+  constructor(inner: EntitiesCatalog[]) {
+    this.inner = inner;
+  }
+
+  async entities(filters?: EntityFilters): Promise<Entity[]> {
+    const ops = this.inner.map(catalog => catalog.entities(filters));
+    const collections = await Promise.all(ops);
+    return collections.flat();
+  }
+
+  entityByUid(uid: string): Promise<Entity | undefined> {
+    return new Promise((resolve, reject) => {
+      const ops = this.inner.map(catalog =>
+        catalog
```

Comment:
One thing i've noticed here is that if one of the inner catalogs fail, it will fail the entire chain rather than just it's own.

If one provider fails, depending on the order of completion, it might never resolve with a working catalog provider. Is this something that we should consider supporting?

Reply:
Maybe something like this might be better?

```ts
entityByUid(uid: string): Promise<Entity | undefined> {
    return new Promise(async (resolve, reject) => {
      const ops = this.inner.map(async (catalog) => {
        try {
          const entity = await catalog.entityByUid(uid);
          if (entity) resolve(entity);
          return entity;
        } catch {
          logger.log("failed collecting from provider");
        }
      });

      const returned = await Promise.all(ops);
      const matched = returned.find(Boolean);
      if (!matched) {
        reject("could not find a matching id");
      }
    });
  }
```

Probably need a little more thought as this is just kind of pseudocode but I don't think that one provider either throwing errors because it can't find an entity or it's unable to process requests should bring down the Coalescing chain,

Reply:
Agreed that non-determinism is bad, question is if a partially working catalog would be any better than if it just went down. I guess it depends on what kinds they have served by what implementation; having users but not components would probably be useless for example and I am worried that masking errors (logging "silently" and not returning the information to the caller) could  be problematic too. Hmmm.

Well given that this is more of an example than a recommended production thing (as per the class comment), let's get your suggestion in and iterate. It does make it deterministic.